# R version 3.6.0

## TODO:
# modify plots to show p-value and change above a line connecting the two relevant normalized count

####  load/install required packages  ####
packages = c("gplots","RColorBrewer","BiocManager","utils", "ggplot2", "ggrepel", "factoextra", "cowplot", "plyr")
package.check <- function(x){
  if (!require(x, character.only = TRUE)) {
    install.packages(x, dependencies = TRUE)
    library(x, character.only = TRUE)
  }
}
for(pkg in packages) {
  package.check(pkg)
}
if(!require('DESeq2',character.only = TRUE)) {
  BiocManager::install("DESeq2")
  library("DESeq2")
}

####  data preprocessing ####
# set working directory to script directory
# This line only works when working in R Studio
setwd(dirname(rstudioapi::getActiveDocumentContext()$path))

# load gene id and gene names
# this file can be found in the same github folder
geneid <- read.csv("gene_ids_ws271.csv")
rowids = geneid$X
geneid <- geneid[,-1]
row.names(geneid) <- rowids

# load raw count data generated by featureCount
# This data can be found in the NIH Gene Expression Omnibus - GSE209988
# This file should be extracted into the same folder as this R file
ocyts <- read.csv("GSE209988_processed_reads_per_gene.csv", comment.char="#", skip=1, stringsAsFactors=FALSE)

# remove Chr, Start, End, Length, Strand attributes
ocyts <- ocyts[,-c(2:6)]

# remove .cutadapt.sam from sample names
colnames(ocyts) <- c('geneid',gsub(".*(m..c.).*","\\1",colnames(ocyts)[2:ncol(ocyts)]))

# save geneid in temporary vector ids
ids <- ocyts$geneid

# delete column of geneid now, format matrix into integer matrix
ocyts <- ocyts[,-1]
ocyts <- apply(ocyts,2,as.integer)

# set geneid as rownames
row.names(ocyts) <- ids

# ocyts with row names
ocytsR <- ocyts

temp <- geneid[row.names(ocytsR),]
row.names(ocytsR) <- temp$gene.name
ocytsR <- cbind(ocytsR, temp$gene.name)

####  basic sanity check  ####
# check read depth for each cell
# conclusion: all cells got >8M reads
count.in.millions <- colSums(ocyts)/1e6

# Figure S1A
par(las=2)
par(mar=c(5,5,5,5))
barplot(count.in.millions, main="Counts per Oocyte", xlab="Oocyte Isolate", ylab="Total Reads (1e6)")

# check if read depth is sufficient to detect all genes
# conclusion: read depth is sufficient, ~8000 genes detected at a threshold of 1 CPM
# calculate CPM
cpm <- t(t(ocyts)/count.in.millions)

# number of gene detected by different threshold
th.100 <- colSums(cpm>100)
th.10 <- colSums(cpm>10)
th.1 <- colSums(cpm>1)
th.0.1 <- colSums(cpm>0.1)

# Figure S1B
temp1 <- data.frame(gene.num=th.100,threshold=100)
rownames(temp1) <- c()
temp2 <- data.frame(gene.num=th.10,threshold=10)
rownames(temp2) <- c()
temp3 <- data.frame(gene.num=th.1,threshold=1)
rownames(temp3) <- c()
temp4 <- data.frame(gene.num=th.0.1,threshold=0.1)
rownames(temp4) <- c()
temp <- rbind(temp1,temp2,temp3,temp4)
par(mfrow=c(1,1))
boxplot(gene.num~threshold,data = temp, xlab='Count Threshold (1e6)', ylab='gene detected', main='')

# check if read depth is sufficient
# when read depth is insufficient, the number of gene detected increase with count, especially for low detection threshold, e.g. 0.1 CPM
par(mfrow=c(1,4))
plot(colSums(ocyts)/1e6,th.0.1,xlab='Total Reads (1e6)',ylab='gene detected',main='0.1 cpm')
plot(colSums(ocyts)/1e6,th.1,xlab='Total Reads (1e6)',ylab='',main='1 cpm')
plot(colSums(ocyts)/1e6,th.10,xlab='Total Reads (1e6)',ylab='',main='10 cpm')
plot(colSums(ocyts)/1e6,th.100,xlab='Total Reads (1e6)',ylab='', main='100 cpm')


####  differential expression analysis ####
## get normalized counts
# basic filtering, keep genes with >2 counts in at least 8 (out of 12) oocytes
cts <- ocyts[rowSums(ocyts>2)>8,]

# setup design matrix for DESeq2
coldata <- as.matrix(substr(colnames(cts),1,2))
colnames(coldata) <- c('condition')
rownames(coldata) <- colnames(cts)
summary(coldata)
# run DESeq2, should take <2 minutes
dds <- DESeqDataSetFromMatrix(countData = cts,
                              colData = coldata,
                              design = ~ condition)
dds <- DESeq(dds)

# extract normalized counts
cd <- counts(dds,normalize=TRUE)
temp <- geneid[row.names(cd),]
row.names(cd) <- temp$gene.name

## do differential expression between -5 and -3
# basic filtering, keep genes with >2 counts in at least 8 (out of 12) oocytes
cts <- ocyts[rowSums(ocyts>2)>8,]
# remove columes containing -3 oocytes
# note: in theory, DESeq2 can take >2 groups for differential expression analysis, but in reality it never works well.
# therefore, we only compare 2 groups at a time. -5 vs -1 first, then -3 vs -1.
cts <- cts[,c(-1,-2,-3,-4)]

# setup design matrix for DESeq2
coldata <- as.matrix(substr(colnames(cts),1,2))
colnames(coldata) <- c('condition')
rownames(coldata) <- colnames(cts)
summary(coldata)
# run DESeq2, should take <2 minutes
dds <- DESeqDataSetFromMatrix(countData = cts,
                              colData = coldata,
                              design = ~ condition)
dds <- DESeq(dds)
# extract results
res <- results(dds)
# get DEGs
# order by adjusted p value
res <- res[order(res$padj),]
res <- as.data.frame(res)
res <- res[complete.cases(res),]
m3.m5.all <- res
rownames(m3.m5.all) <- geneid[rownames(m3.m5.all),]$gene.name
# only keep significantly upregulated genes in -1
keep <- (res$padj<0.05) & (res$log2FoldChange < -1)
# note: these DEGs are selected only based on -5 and -1
m3.m5.degs <- res[keep,]
m3.m5.degs <- m3.m5.degs[complete.cases(m3.m5.degs),]
rownames(m3.m5.degs) <- geneid[rownames(m3.m5.degs),]$gene.name
m3.m5.degs$test <- '-3 vs -5'


## do differential expression between -5 and -1
# basic filtering, keep genes with >2 counts in at least 8 (out of 12) oocytes
cts <- ocyts[rowSums(ocyts>2)>8,]
# remove columes containing -3 oocytes
# note: on theory, DESeq2 can take >2 groups for differential expression analysis, but in fact it never works well.
# therefore, we only compare 2 groups at a time. -5 vs -1 first, then -3 vs -1.
cts <- cts[,c(-5,-6,-7,-8)]
# setup design matrix for DESeq2
coldata <- as.matrix(substr(colnames(cts),1,2))
colnames(coldata) <- c('condition')
rownames(coldata) <- colnames(cts)
summary(coldata)
# run DESeq2, should take <2 minutes
dds <- DESeqDataSetFromMatrix(countData = cts,
                              colData = coldata,
                              design = ~ condition)
dds <- DESeq(dds)
# extract results
res <- results(dds)
# get DEGs
# order by adjusted p value
res <- res[order(res$padj),]
res <- as.data.frame(res)
res <- res[complete.cases(res),]
m1.m5.all <- res
rownames(m1.m5.all) <- geneid[rownames(m1.m5.all),]$gene.name
# only keep significantly upregulated genes in -1
keep <- (res$padj<0.05) & (res$log2FoldChange < -1)
# note: these DEGs are selected only based on -5 and -1
m1.m5.degs <- res[keep,]
m1.m5.degs <- m1.m5.degs[complete.cases(m1.m5.degs),]
rownames(m1.m5.degs) <- geneid[rownames(m1.m5.degs),]$gene.name
m1.m5.degs$test <- '-1 vs -5'

## m1.m5.degs contain genes upregulated in -1 compared to -5
## there could also be genes upregulated in -1 compared to -3, but have similar levels between -5 and -3
## then, do differential expression between -3 and -1
# basic filtering, keep genes with >2 counts in at least 8 (out of 12) oocytes
cts <- ocyts[rowSums(ocyts>2)>8,]
# remove column containing -5 oocytes
cts <- cts[,c(-9,-10,-11,-12)]
# setup design matrix for DESeq2
coldata <- as.matrix(substr(colnames(cts),1,2))
colnames(coldata) <- c('condition')
rownames(coldata) <- colnames(cts)
summary(coldata)
# run DESeq2, should take <2 minutes
dds <- DESeqDataSetFromMatrix(countData = cts,
                              colData = coldata,
                              design = ~ condition)
dds <- DESeq(dds)
# extract results
res <- results(dds)
# get DEGs
# order by adjusted p value
res <- res[order(res$padj),]
res <- as.data.frame(res)
res <- res[complete.cases(res),]
m1.m3.all <- res
rownames(m1.m3.all) <- geneid[rownames(m1.m3.all),]$gene.name
# only keep significantly upregulated genes in -1
keep <- (res$padj<0.05) & (res$log2FoldChange < -1)
# note: these DEGs are selected only based on -3 and -1
m1.m3.degs <- res[keep,]
m1.m3.degs <- m1.m3.degs[complete.cases(m1.m3.degs),]
rownames(m1.m3.degs) <- geneid[rownames(m1.m3.degs),]$gene.name
m1.m3.degs$test <- '-1 vs -3'


#### figure style visualizations (Figure 2A)
### volcano plot of each comparison (m3 or m5 to m1)

## plot for m1.m3.all.
# Since a negative fold change reflects an increase when going from -3 -> -1,
# Negate all of log2FoldChange so that positive reflects an increase when going from -3 -> -1
m1.m3.all$log2FoldChange <- -(m1.m3.all$log2FoldChange)
#Add column for differential expression
m1.m3.all$diffexpressed <- 'NO'
#if log2FoldChange > 1, pvalue < 0.05 set as 'up'
m1.m3.all$diffexpressed[m1.m3.all$log2FoldChange > 1 & m1.m3.all$padj < 0.05] <- "UP"
#if log2FoldChange < -1, pvalue < 0.05 set as 'down'
m1.m3.all$diffexpressed[m1.m3.all$log2FoldChange < -1 & m1.m3.all$padj < 0.05] <- "DOWN"

# Set a label column
m1.m3.all$delabel <- NA
m1.m3.all$delabel[m1.m3.all$diffexpressed != "NO"] <- row.names(m1.m3.all)[m1.m3.all$diffexpressed != "NO"]
# make initial volcano plot
m3plot <- ggplot(data=m1.m3.all, aes(x=log2FoldChange, y=-log10(padj), col=diffexpressed, label=delabel)) + geom_point(size=2)
# print gene names as text
m3plot <- m3plot + theme_bw() + theme(legend.position = "none")
# set lines for volcano plot
m3plot <- m3plot + geom_vline(xintercept=c(-1, 1), linetype="dashed") + geom_hline(yintercept=-log10(0.05), linetype="dashed")
# set colors for volcano plot ("Down", "No", "UP")
m3plot <- m3plot + scale_color_manual(values=c("blue", "black", "red")) + labs(title='-1 (N=4) vs -3 (N=4)') + xlab("Fold Change (log2)") + ylab("p-value(-log10)")
# set axis scale numbers
m3plot <- m3plot + scale_y_continuous(breaks=seq(0,4,by=1)) + scale_x_continuous(breaks=seq(-6,6,by=1))


## Do it again for m1.m5
# Negate all of log2FoldChange so that positive reflects an increase when going from -5 -> -1
m1.m5.all$log2FoldChange <- -(m1.m5.all$log2FoldChange)
#Add column for differential expression
m1.m5.all$diffexpressed <- 'NO'
#if log2FoldChange > 1, pvalue < 0.05 set as 'up'
m1.m5.all$diffexpressed[m1.m5.all$log2FoldChange > 1 & m1.m5.all$padj < 0.05] <- "UP"
#if log2FoldChange < -1, pvalue < 0.05 set as 'down'
m1.m5.all$diffexpressed[m1.m5.all$log2FoldChange < -1 & m1.m5.all$padj < 0.05] <- "DOWN"
# Set a label column
m1.m5.all$delabel <- NA
m1.m5.all$delabel[m1.m5.all$diffexpressed != "NO"] <- row.names(m1.m5.all)[m1.m5.all$diffexpressed != "NO"]

# setup the volcano plot
m5plot <- ggplot(data=m1.m5.all, aes(x=log2FoldChange, y=-log10(padj), col=diffexpressed, label=delabel)) + geom_point(size=2)
# print gene names as text
m5plot <- m5plot + theme_bw()
# set lines for volcano plot
m5plot <- m5plot + geom_vline(xintercept=c(-1, 1), linetype="dashed") + geom_hline(yintercept=-log10(0.05), linetype="dashed")
# set colors for volcano plot ("Down", "No", "UP")
m5plot <- m5plot + scale_color_manual(values=c("blue", "black", "red")) + labs(title="-1 (N=4) vs -5 (N=4)") + xlab("Fold Change (log2)") + ylab("")
# set axis scale numbers
m5plot <- m5plot + scale_y_continuous(breaks=seq(0,20,by=1)) + scale_x_continuous(breaks=seq(-6,6,by=1))

# Arrange and display figure 2a
figure2a <- plot_grid(m3plot, m5plot, rel_widths = c(11, 16))
figure2a

## Table S1
# Separate out UP and DOWN expression data for each group
m1.m3.up <- m1.m3.all[m1.m3.all$diffexpressed == 'UP',]
m1.m3.down <- m1.m3.all[m1.m3.all$diffexpressed == 'DOWN',]

m1.m5.up <- m1.m5.all[m1.m5.all$diffexpressed == 'UP',]
m1.m5.down <- m1.m5.all[m1.m5.all$diffexpressed == 'DOWN',]

cols_to_grab <- c("baseMean", "log2FoldChange", "padj")

up.intersect <- intersect(row.names(m1.m3.up), row.names(m1.m5.up))
m3.up.unique <- row.names(m1.m3.up[setdiff(row.names(m1.m3.up), up.intersect),])
m5.up.unique <- row.names(m1.m5.up[setdiff(row.names(m1.m5.up), up.intersect),])

class1.m3 <- m1.m3.up[m3.up.unique, cols_to_grab]
colnames(class1.m3) <- c("m3_baseMean", "m3_log2FoldChange", "m3_padj")

class1.m5 <- m1.m5.up[m5.up.unique, cols_to_grab]
colnames(class1.m5) <- c("m5_baseMean", "m5_log2FoldChange", "m5_padj")

class1.bot <- rbind.fill(class1.m3, class1.m5)
rownames(class1.bot) <- c(m3.up.unique, m5.up.unique)

class1 <- m1.m3.up[up.intersect, cols_to_grab]
colnames(class1) <- c("m3_baseMean", "m3_log2FoldChange", "m3_padj")
class1 <- cbind(class1, m1.m3.up[up.intersect, cols_to_grab])
colnames(class1) <- c("m3_baseMean", "m3_log2FoldChange", "m3_padj", "m5_baseMean", "m5_log2FoldChange", "m5_padj")

class1 <- rbind(class1, class1.bot)

# Write class 1 results to file
write.csv(class1, 'class1.csv', row.names=TRUE, na="---")

# Generate table for class 2 transcripts
down.intersect <- intersect(row.names(m1.m3.down), row.names(m1.m5.down))
m3.down.unique <- row.names(m1.m3.down[setdiff(row.names(m1.m3.down), down.intersect),])
m5.down.unique <- row.names(m1.m5.down[setdiff(row.names(m1.m5.down), down.intersect),])

class2.m3 <- m1.m3.down[m3.down.unique, cols_to_grab]
colnames(class2.m3) <- c("m3_baseMean", "m3_log2FoldChange", "m3_padj")

class2.m5 <- m1.m5.down[m5.down.unique, cols_to_grab]
colnames(class2.m5) <- c("m5_baseMean", "m5_log2FoldChange", "m5_padj")

class2.bot <- rbind.fill(class2.m3, class2.m5)
rownames(class2.bot) <- c(m3.down.unique, m5.down.unique)

class2 <- m1.m3.down[down.intersect, cols_to_grab]
colnames(class2) <- c("m3_baseMean", "m3_log2FoldChange", "m3_padj")
class2 <- cbind(class2, m1.m3.down[down.intersect, cols_to_grab])
colnames(class2) <- c("m3_baseMean", "m3_log2FoldChange", "m3_padj", "m5_baseMean", "m5_log2FoldChange", "m5_padj")

class2 <- rbind(class2, class2.bot)

# Write class 2 results to file
write.csv(class2, 'class2.csv', row.names=TRUE, na="---")
